Python/Product/Analysis/Interpreter/Ast/AstAnalysisWalker.cs

```diff
@@ -0,0 +1,260 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Parsing;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Interpreter.Ast {
+    class AstAnalysisWalker : PythonWalker {
+        private readonly IPythonInterpreter _interpreter;
+        private readonly PythonAst _ast;
+        private readonly IPythonModule _module;
+        private readonly string _filePath;
+        private readonly Dictionary<string, IMember> _members;
+
+        private readonly Stack<Dictionary<string, IMember>> _scope;
+
+        private IMember _noneInst;
+
+        public AstAnalysisWalker(
+            IPythonInterpreter interpreter,
+            PythonAst ast,
+            IPythonModule module,
+            string filePath,
+            Dictionary<string, IMember> members
+        ) {
+            _interpreter = interpreter;
+            _ast = ast;
+            _module = module;
+            _filePath = filePath;
+            _members = members;
+            _scope = new Stack<Dictionary<string, IMember>>();
+            _noneInst = new AstPythonConstant(_interpreter.GetBuiltinType(BuiltinTypeId.NoneType));
+        }
+
+        public override bool Walk(PythonAst node) {
+            if (_ast != node) {
+                throw new InvalidOperationException("walking wrong AST");
+            }
+            _scope.Push(_members);
+            return base.Walk(node);
+        }
+
+        public override void PostWalk(PythonAst node) {
+            _scope.Pop();
+            base.PostWalk(node);
+        }
+
+        internal LocationInfo GetLoc(Node node) {
+            if (node == null || node.StartIndex >= node.EndIndex) {
+                return null;
+            }
+
+            var start = node.GetStart(_ast);
+            var end = node.GetEnd(_ast);
+            return new LocationInfo(_filePath, start.Line, start.Column, end.Line, end.Column);
+        }
+
+        private IMember GetValueFromExpression(Expression expr, Dictionary<string, IMember> scope) {
```

Comment:
I remember planning to do that... not sure why I didn't. As it happens, the only relevant scopes are the top level and the current class, since we don't go into functions at all. I doubt there are all that many cases where a module-level variable is cloned into a class variable, but it's easy enough to check the module scope here as well.

---
In reply to: [83940069](https://github.com/Microsoft/PTVS/pull/1729#discussion_r83940069) [](ancestors = 83940069)